docs(kinesis): refresh README and add updated snippets (#7038)
aws/aws-cdk: packages/@aws-cdk/aws-kinesis/README.md

---
<!--END STABILITY BANNER-->

> [Amazon Kinesis](https://docs.aws.amazon.com/streams/latest/dev/introduction.html) provides collection and processing of large [streams](https://aws.amazon.com/streaming-data/) of data records in real time. Kinesis data streams can be used for rapid and continuous data intake and aggregation.

Reviewer: line width to 120 or 150 throughout.

> ## Table Of Contents

Reviewer: Usually nice to leave one empty line before and after a header.

```suggestion
## Table Of Contents

```

- [Streams](#streams)
  - [Encryption](#encryption)
  - [Import](#import)

> ### Streams

Reviewer: Use h2 instead of h3. Or, switch 'table to contents' to h3 as well.

I would suggest the former.

Author: done

Amazon Kinesis Data Streams ingests a large amount of data in real time, durably stores the data, and makes the data available for consumption.

Using the CDK, a new Kinesis stream can be created as part of the stack using the construct's constructor. You may specify the `streamName` to give your own identifier to the stream. If not, CloudFormation will generate a name. 

```ts
new Stream(this, 'MyFirstStream', {
    streamName: 'my-awesome-stream',
});
```

You can also specify properties such as `shardCount` to indicate how many shards the stream should choose and a `retentionPeriod`

> to specify how many hours data in shards should remain accessible. 

Reviewer: ```suggestion
to specify how long the data in the shards should remain accessible. Read more at [Creating and Managing 
Streams](https://docs.aws.amazon.com/streams/latest/dev/working-with-streams.html)
```

Author: done

```ts
new Stream(this, 'MyFirstStream', {
    streamName: 'my-awesome-stream',
    shardCount: 3,
    retentionPeriod: Duration.hours(48)
});
```

Streams are not encrypted by default.

#### Encryption

[Stream encryption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-kinesis-stream-streamencryption.html) enables server-side encryption using an AWS KMS key for a specified stream.

You can enable encryption on your streams by specifying the `encryption` property. A KMS key will be created for you and associated with the stream. 

```ts
const stream = new Stream(this, 'MyEncryptedStream', {
    encryption: StreamEncryption.KMS
});
```

> You can also provide your own external KMS key to use for stream encryption by specifying the `encryptionKey` property.

Reviewer: ```suggestion
You can also supply your own external KMS key to use for stream encryption by specifying the `encryptionKey` property.
```

```ts
import * as kms from '@aws-cdk/aws-kms';

const key = new kms.Key(this, 'MyKey');

const stream = new Stream(this, 'MyEncryptedStream', {
    encryption: StreamEncryption.KMS,
    encryptionKey: key
});
```

#### Import

Any Kinesis stream that has been created outside the stack can be imported into your CDK app.

> Streams can be imported by their ARN via the `Stream.fromStreamArn` API

Reviewer: ```suggestion
Streams can be imported by their ARN via the `Stream.fromStreamArn()` API
```

Author: done

```ts
const stack = new Stack (app, 'MyStack');

const importedStream = Stream.fromStreamArn(stack, 'ImportedStream', 'arn:aws:kinesis:us-east-2:123456789012:stream/f3j09j2230j')
```

> Encrypted Streams can also be imported by their attributes via the `Stream.fromStreamAttributes` API

Reviewer: Do we need a separate API here? Can't we just extend the `fromStreamArn()` API with an optional kmsKey attribute?

Author: I think that's a good point. I'm not sure this PR is the right place to make that decision, but let's discuss it further.

It would be a breaking change but also a simplified user experience. Absolutely worth considering.
